i2c/imx6ull/libimx6ull-i2c.c

```diff
@@ -0,0 +1,368 @@
+	/* generate STOP: disable master mode, disable TX */
+	*(i2c.base + i2cr) &= ~((1 << 6) | (1 << 5) | (1 << 4));
+
+	/* wait for the bus to be idle */
+	waitBusBusy(0); /* don't overwrite possible error in ret */
+
+	return ret;
+}
+
+
+static int _doRead(uint8_t dev_addr, uint8_t *data_out, uint32_t len)
+{
+	unsigned int i;
+	uint16_t temp;
+	int ret;
+
+	/* send first byte - DEVICE READ request */
+	ret = writeByte((dev_addr << 1) | i2c_cmd_read);
+	if (ret < 0)
+		return ret;
+
+	/* setup bus for data readout - disable TX, enable ACKing for every byte except the last one */
+	temp = *(i2c.base + i2cr) & ~((1 << 4) | (1 << 3));
+	if (len <= 1)
+		temp |= (1 << 3);
+	*(i2c.base + i2cr) = temp;
+
+	/* RM: dummy read to enter read mode */
+	(void)*(i2c.base + i2dr);
+
+	/* read data */
+	for (i = 0; (ret == 0) && (i < len); ++i) {
+		ret = i2c_trxComplete();
+		if (ret < 0)
+			break;
+
+		if (i == len - 1) {
+			/* we must generate STOP before reading i2dr to avoid generating another clock cycle */
+			*(i2c.base + i2cr) &= ~((1 << 5) | (1 << 4));
+		}
+		else if (i == len - 2) {
+			/* we must set TXAK before receiving next-to-last byte to acknowledge the whole message */
+			*(i2c.base + i2cr) |= (1 << 3);
+		}
+
+		data_out[i] = *(i2c.base + i2dr);
+	}
+
+	return ret;
+}
+
+
+/* Performs i2c generic read operation from the given slave device. */
+int i2c_busRead(uint8_t dev_addr, uint8_t *data_out, uint32_t len)
+{
+	int ret;
+
+	/* single-master mode, bus should always be idle when we're not transferring */
+	if (isBusBusy())
+		return -EBUSY;
+
+	/* generate START: enable master mode, enable TX for sending register address */
+	*(i2c.base + i2cr) |= (1 << 6) | (1 << 5) | (1 << 4);
+
+	/* wait for the bus to be busy */
+	ret = waitBusBusy(1);
+
+	if (ret >= 0)
+		ret = _doRead(dev_addr, data_out, len);
+
+	/* generate STOP: disable master mode, disable TX (might be done twice without any issue) */
+	*(i2c.base + i2cr) &= ~((1 << 6) | (1 << 5) | (1 << 4));
+
+	/* wait for the bus to be idle */
+	waitBusBusy(0); /* don't overwrite possible error in ret */
+
+	return ret;
+}
+
+
+/* Performs i2c regiester read operation from the given slave device */
+int i2c_regRead(uint8_t dev_addr, uint8_t reg_addr, uint8_t *data_out, uint32_t len)
+{
+	int ret;
+
+	ret = i2c_busWrite(dev_addr, &reg_addr, 1);
+	if (ret < 0)
+		return ret;
+
+	return i2c_busRead(dev_addr, data_out, len);
+}
+
+
+int i2c_init(unsigned int dev_no)
+{
+	i2c.dev_no = dev_no;
+
+	if ((i2c.dev_no < 1) || (i2c.dev_no > (sizeof(i2c_addr) / sizeof(i2c_addr[0]))))
+		return -1;
+
+	/* initialize */
+	i2c.base = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, i2c_addr[i2c.dev_no - 1]);
+
+	if (i2c.base == MAP_FAILED)
+		return -2;
+
+	platform_init();
+
+	/* prepare interrupt routine */
+	if (mutexCreate(&i2c.lock) != EOK) {
+		munmap((void *)i2c.base, _PAGE_SIZE);
+		return -3;
+	}
+
+	if (condCreate(&i2c.cond) != EOK) {
+		munmap((void *)i2c.base, _PAGE_SIZE);
```

Comment:
Missing `resourceDestroy(i2c.lock);`

Reply:
From what I understand this driver can be used as lib - so failure here not always cause process termination

Reply:
thx, done.
Yeah, you're right it might be used that way (I've focused only on my use cases - so i2c device drivers exit if they cannot initialize i2c bus). Added also `i2c.initialized` to avoid accessing `i2c.base` if it's not mapped correctly and somebody would not check `i2c_init` return value.